File: GooseClient/src/sample/mina/MinaHelper.java
package sample.mina;

import com.google.gson.Gson;
import org.apache.mina.core.filterchain.DefaultIoFilterChainBuilder;
import org.apache.mina.core.future.ConnectFuture;
import org.apache.mina.core.session.IoSession;
import org.apache.mina.filter.codec.ProtocolCodecFilter;
import org.apache.mina.transport.socket.nio.NioSocketConnector;

import java.net.InetSocketAddress;

/**
 * @author DeMon
 * Created on 2020/5/19.
 * E-mail 757454343@qq.com
 * Desc:
 */
public class MinaHelper {
    private static class Helper {
        private static MinaHelper INSTANCE = new MinaHelper();
    }

    private Gson gson;

    private MinaHelper() {
        //构造函数
        gson = new Gson();
    }

    public static MinaHelper getInstance() {
        return Helper.INSTANCE;
    }

    private IoSession session;

    public void start(String ip, int port, MinaCallback callback) {
        if (session != null && session.isConnected()) {
            return;
        }
        try {
            // 创建一个非阻塞的server端的Socket
            NioSocketConnector connector = new NioSocketConnector();
            // 设置过滤器（使用mina提供的文本换行符编解码器）
            DefaultIoFilterChainBuilder filterChain = connector.getFilterChain();
            filterChain.addLast("codec", new ProtocolCodecFilter(new MsgCodecFactory()));
            // 为接收器设置管理服务
            connector.setHandler(new ServiceHandler(callback));
            ConnectFuture future = connector.connect(new InetSocketAddress(ip, port));
            future.awaitUninterruptibly();
            session = future.getSession();
        } catch (Exception e) {
            e.printStackTrace();
            callback.bindException();
        }
    }

    public void stop() {
        if (session != null && session.isConnected()) {
            session.close(true);
            session = null;
        }
    }

    public boolean isConnected(){
        return session != null && session.isConnected();
    }


    public void sendMsg(String msg) {
        sendMsg(new Message(msg));
    }


    public void sendMsg(Message msg) {
        if (session == null || !session.isConnected()) {
            return;
        }
        session.write(gson.toJson(msg));
    }

}

File: MinaApp/app/src/main/java/com/demon/imapp/mina/MinaHelper.kt
package com.demon.imapp.mina

import com.demon.imapp.data.Const
import com.demon.imapp.data.Message
import com.demon.imapp.util.SPUtil
import com.demon.imapp.util.toJson
import kotlinx.coroutines.*
import org.apache.mina.core.session.IdleStatus
import org.apache.mina.core.session.IoSession
import org.apache.mina.filter.codec.ProtocolCodecFilter
import org.apache.mina.filter.keepalive.KeepAliveFilter
import org.apache.mina.filter.keepalive.KeepAliveRequestTimeoutHandler
import org.apache.mina.transport.socket.nio.NioSocketConnector
import java.net.InetSocketAddress

/**
 * @author DeMon
 * Created on 2020/5/15.
 * E-mail 757454343@qq.com
 * Desc:
 */
object MinaHelper {
    private val coroutineScopeIO = CoroutineScope((SupervisorJob() + Dispatchers.IO))

    private const val timeout = 10 * 1000L

    private var session: IoSession? = null

    /**
     * 连接job
     */
    private var connectJob: Job? = null

    fun connect() {
        val ip = SPUtil.getInstance().get(Const.IP, "") as String
        val port = SPUtil.getInstance().get(Const.PORT, "") as String
        val name = SPUtil.getInstance().get(Const.NAME, "") as String
        connectJob = coroutineScopeIO.launch {
            val connector = NioSocketConnector()
            //设置链接超时时间
            connector.connectTimeoutMillis = timeout
            //添加包编码过滤器 ，处理编码结束符
            connector.filterChain.addLast("codec", ProtocolCodecFilter(MsgCodecFactory()))
            connector.handler = MinaReceiver()
            connector.addListener(MinaListener())
         /*   //设置心跳包
            val heartFilter = KeepAliveFilter(KeepAliveFactory(), IdleStatus.READER_IDLE, KeepAliveRequestTimeoutHandler.CLOSE)
            heartFilter.isForwardEvent = true
            //发送一个心跳包
            heartFilter.requestInterval = 10
            //心跳包超时时间 10s
            heartFilter.requestTimeout = 10
            connector.filterChain.addLast("heartbeat", heartFilter)*/
            runCatching {
                val future = connector.connect(InetSocketAddress(ip, port.toInt()))
                // 等待连接创建完成
                future.awaitUninterruptibly()
                //获得session
                session = future.session
            }.onFailure {
                connector.dispose()
            }
        }
    }

    fun getSessionId(): Long? {
        return session?.id
    }

    fun disConnect() {
        connectJob?.cancel()
        session?.close(true)
    }


    fun isConnected(): Boolean {
        session?.run {
            return isConnected
        }
        return false
    }

    fun sendMsg(msg: String) {
        sendMsg(Message(msg))
    }

    fun sendMsg(msg: Message) {
        session?.write(msg.toJson())
    }


}